src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetLockService.cs

```diff
@@ -44,42 +44,47 @@ public NuGetLockService(JoinableTaskContext joinableTaskContext)
         /// </summary>
         public async Task<T> ExecuteNuGetOperationAsync<T>(Func<Task<T>> action, CancellationToken token)
         {
-            return await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            if (_lockCount.Value == 0)
```

Comment:
Unrelated to this change.  It is kind of interesting pattern to use AsyncLocal to track the task chain holds an existing semaphore.  Generally, I don't think it is a safe pattern, because the ExecuteContext is extremely likely to be leaked, and reused later, when the main task is done, and semaphore is released.  I think you would want to change that to carry a signature of the Semaphore holder, and compare it with the current semaphore holder to know whether it is still the owner of it.  But even with that, the main owner can release the lock, while a spin off task is still using that.  Of course that is an unrelated issue, so I suggest to think about it more later.